### PR TITLE
Add `BatchSizeSchedulerCallback`

### PR DESCRIPTION
**Pull request by epwalsh**
Implements the batch size scheduler defined by Dirk in [this document](https://docs.google.com/document/d/1D_LCmHdJcTAVsiLveDdik1F2VwuSlsnm3A9ulnNXQA8/edit?usp=sharing).

### CODE REVIEW

File: src/olmo_core/train/callbacks/batch_size_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                elif target_batch_size < self.current_batch_size:
+                    assert self.current_batch_size % target_batch_size == 0
+                    self._update_batch_size_and_lr(target_batch_size)
+
+                break
+
+    def _update_batch_size_and_lr(self, batch_size: int):
+        log.info(f"Changing global batch size to {batch_size:,d}...")
+        ratio = batch_size / self.current_batch_size
+        lr_adjustment_factor = math.sqrt(ratio)
```
Comment (dirkgr) [inline src/olmo_core/train/callbacks/batch_size_scheduler.py:124]: It's worth noting that this is true for Adam, but may not be true for other optimizers. I don't know how to package that fact in the code. Can we check somehow whether we have Adam, and throw an error if we don't?

Comment (epwalsh) [reply src/olmo_core/train/callbacks/batch_size_scheduler.py:124]: 29a9ae02

**Comment by epwalsh** [comment]
I have some short test runs here: https://www.comet.com/ai2/petew-test-bz-sched/view/new/panels

File: src/olmo_core/train/callbacks/batch_size_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    def __post_init__(self):
+        if len(self.batch_sizes) != len(self.schedule):
+            raise OLMoConfigurationError(
+                "batch_sizes and schedules should have the same number of items"
+            )
+
+        if not self.schedule:
+            return
+
+        if len(set([duration.unit for duration in self.schedule])) > 1:
```
Comment (dirkgr) [inline src/olmo_core/train/callbacks/batch_size_scheduler.py]: Python has set comprehensions too. `{x for x in something}`.

File: src/olmo_core/train/callbacks/batch_size_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                f"Unable to adjust learning rate for {self.trainer.train_module.__class__.__name__} train module class"
+            )
+            return
+
+        for optim in optimizers:
+            if not isinstance(optim, (torch.optim.Adam, torch.optim.AdamW, SkipStepAdamW)):
+                log.warning(
+                    f"Unable to adjust learning rate for {optim.__class__.__name__} optimizer"
+                )
+                return
```
Comment (dirkgr) [inline src/olmo_core/train/callbacks/batch_size_scheduler.py]: It falls back to doing nothing? I'd rather it fails. Otherwise we end up running 500B tokens worth of baselines, only to discover afterwards that something we configured didn't actually run.

File: src/olmo_core/train/callbacks/batch_size_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+                if group.get(initial_lr_field) is None:
+                    group[initial_lr_field] = group[lr_field]
+
+                group[initial_lr_field] *= lr_adjustment_factor
+
+                # Only set the actual LR if there's no scheduler. Schedulers update the LR based
+                # on the initial LR.
+                if scheduler is None:
+                    group[lr_field] = group[initial_lr_field]
```
Comment (dirkgr) [inline src/olmo_core/train/callbacks/batch_size_scheduler.py:165]: LR schedulers don't overwrite this every step?

**Review by dirkgr — CHANGES_REQUESTED** [review]
What LR will we log? The one from the LR scheduler, or the overall one?

File: src/olmo_core/train/callbacks/batch_size_scheduler.py (hunk 4 above)
Comment (epwalsh) [reply src/olmo_core/train/callbacks/batch_size_scheduler.py:165]: They overwrite "lr" but not "initial_lr". So we only touch "lr" when there's no scheduler.

File: src/olmo_core/train/callbacks/batch_size_scheduler.py (hunk 3 above)
Comment (epwalsh) [reply src/olmo_core/train/callbacks/batch_size_scheduler.py]: done: ff895e8f

File: src/olmo_core/train/callbacks/batch_size_scheduler.py (hunk 2 above)
Comment (epwalsh) [reply src/olmo_core/train/callbacks/batch_size_scheduler.py]: done: ff895e8f

**Comment by epwalsh** [comment]
> What LR will we log? The one from the LR scheduler, or the overall one?

We always log (to comet/W&B) the LR that's actually used in the optimizer step.

**Review by dirkgr — APPROVED** [review]
